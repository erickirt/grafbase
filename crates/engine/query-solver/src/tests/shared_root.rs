use crate::assert_solving_snapshots;

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  nested: Nested!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Nested
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[tokio::test]
async fn all_fields() {
    assert_solving_snapshots!(
        "all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
        "#
    );
}

#[tokio::test]
async fn nested_id_field_with_all_fields() {
    assert_solving_snapshots!(
        "nested_id_with_all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            name
            nested { id }
          }
        }
        "#
    );
}

#[tokio::test]
async fn single_field() {
    assert_solving_snapshots!(
        "single_field",
        SCHEMA,
        r#"
        query {
          products {
            price
          }
        }
        "#
    );
}
