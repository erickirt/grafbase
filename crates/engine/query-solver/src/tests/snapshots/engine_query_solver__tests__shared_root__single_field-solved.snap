---
source: crates/engine/query-solver/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Product.price\", color=forestgreen ]\n    2 [ label = \"Root#price\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    3 [ label = \"products#price\", shape=box, color=dodgerblue, color=forestgreen ]\n    4 [ label = \"price#price\", shape=box, color=dodgerblue, color=forestgreen ]\n    0 -> 2 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    2 -> 3 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    3 -> 4 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 1 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Product.price", steiner=1 ]
    2 [ label = "Root#price", steiner=1 ]
    3 [ label = "products#price", steiner=1 ]
    4 [ label = "price#price", steiner=1 ]
    0 -> 2 [ cost=1, steiner=1]
    2 -> 3 [ cost=0, steiner=1]
    3 -> 4 [ cost=0, steiner=1]
    4 -> 1 [ cost=0, steiner=1]
}
